### DISCUSSION TITLE
Modeling relationships between many types

### DISCUSSION THREAD

**Discussion by divvan** [ravendb/ravendb#19357] [Q&A]
It's common to have many-to-many relationships between two entity types, such as users and groups. But what do you do when one type can reference many other types?

I work in factories where the machinery has many different types of entities, such as machines, sensors and computers. Because they can all develop problems, I also have a Problem entity that can be associated with any one of them.

There are reasons why I can't record that association on the machinery entities. Also, each Problem is normally related to only one entity, whereas a machine may have dozens of problems over it lifetime.

So with C#, I'm thinking of having a MachineryIds array in the Problem class that's defined on the base class for the machinery entities. Is that a valid way of doing it with RavenDB, or is there a better way?

**Comment by lukaszdobrzynski** [comment]
Let's invert the dependencies and allow your system to report each problem as a separate document. One of the solutions would be to use one-to-many relationship between a `Machine` and a `Problem`. A single machine can have multiple problems. In that case you need a `Problem` entity which holds a reference to the `Id` of a `Machine`. The `Machine` class is the base class all your concrete machines will inherit from plus it knows nothing about its problems:
```
public class Problem
{
     public string MachineId { get; set; }
}

public abstract class Machine 
{
     public abstract string MachineTypeName { get; }
}
public class Sensor : Machine {}
public class Computer : Machine {}

```

You can store all different types of your machines in a single collection using `DocumentConventions` available in the `DocumentStore` object:

```
private static void ConfigureCollections(DocumentConventions conventions)
{
     var defaultFindCollectionName = conventions.FindCollectionName;

     conventions.FindCollectionName = type =>
     {
         if (typeof(Machine).IsAssignableFrom(type))
             return "Machines";

         return defaultFindCollectionName(type);
     };
}
```
If you want to get all problems of a given machine you can do it the following way:
```
session.Query<Problem>().Where(x => x.MachineId == "Machines/2-A")
```

On the other hand, if you would like to get a machine affected by a particular problem you can do it like this:

```
var machine = session.Load<Machine>(problem.MachineId)
```

For further reference on document modelling in RavenDB please see [this chapter](https://ravendb.net/learn/inside-ravendb-book/reader/4.0/3-document-modeling) from `Inside RavenDB` book.

**Reply by divvan** [reply]
Thanks for your reply, Lukasz. But as I said in my original post, there are reasons why I can't record the association on the machine side (which I can't explain here). And although I said "each Problem is _normally_ related to only one entity", that isn't always the case. Sometimes, you'll have the same problem in every instance of a particular entity (usually caused by a common software bug).

The Many to Many Relationships section of the Inside RavenDB book says, "In the context of many-to-many associations, we always place the record of the association on the smaller side". In this case, a Problem usually references one machine (although it may reference many), but a machine usually has lots of problems. So that's why I'm thinking of having a MachineryIds array in the Problem class.

That section of the book shows queries between grandparents and grandchildren. Using those example, I could see how to find all the Problems for a particular device, machine, or computer, but I didn't know how to query the other way. But it looks like the solution is to use DocumentConventions to store those 3 types in the same collection?

**Reply by lukaszdobrzynski** [reply]
Unless your domain requires specific machine types when querying related problems, using an abstract class seems justified. 
If you need to filter certain machine types, you can use the `MachineTypeName` field, as suggested above. 
However, if you need to work with a child class instead of the parent class at some point in the processing pipeline, downcasting will likely become unavoidable.

To avoid inheritance and downcasting, you can alternatively choose not to introduce an abstract `Machine` class and instead design your `Problem` entity as follows:
```
public class Problem 
{
    public List<string> ComputerIds { get; set; }
    public List<string> SensorIds { get; set; }
}
```
Whenever a computer or a sensor is affected by a certain problem, you can store its ID in either the ComputerIds or SensorIds collection.
To retrieve your affected computers or sensors using their ids, you can simply use the `Load` method: 

```
var computers = session.Load<Computer>(problem.ComputerIds);
var sensors = session.Load<Sensor>(problem.SensorIds);
```

To further optimize your `Load` queries and reduce the number of remote calls please read about [performing requests lazily](https://ravendb.net/docs/article-page/6.2/csharp/client-api/session/how-to/perform-operations-lazily) in the RavenDB docs.

**Reply by divvan** [reply]
With most databases, I think the normal approach would be to use a separate array for every type that a Problem references, as you suggested. However, a Problem will eventually reference up 30 different types, and not just the 3 I've mentioned here. 

I think this would make the code more complex in a couple of ways. First, in the Problem class, I'd have to search through 30 arrays to find all the entities each Problem references. And in those 30 entity classes, each one would need a different method to get all of its Problems, instead of using a common method that's inherited from the base class. So, I think it's better to use inheritance and casting in this case.

But I am wondering what effect it will have on indexing to put 30 classes into the same collection? Maybe it will improve the performance a little as the indexing on inherited base class fields won't need to be repeated in 30 different collections? But as those entities are arranged in a hierarchical structure, maybe it will be worse for hierarchical queries?

(By the way, I think this would be an ideal application for a graph database, where it is common to have edges to different types of nodes. Unfortunately, RavenDB deprecated its graph features in version 6. And most graph databases don't have RavenDB's other capabilities)

**Reply by ayende** [reply]
There is a *big* distinction between Class and Collection, mind.

RavenDB doesn't care about the class, only what collection you want to put it on. Usually they are the same, but not always.
It's pretty common to put an inheritance hierarchy in a single collection. And yes, a single collection being indexed is far better than 30.

Note that there is issue with having references to different types, no sure what you mean here.

**Comment by ayende** [comment]
Yes, the idea is that you'll have a `Problem.MachineIds` set of ids. 
Or maybe you have a more complex association, such as problem was found at X, routed to Y, root cause is Z, etc.
Not sure if you need to actually capture that information or not. 

But in short, yes, `MachineIds` is the way to go.

The common way to record many-to-many associations in RavenDB is using an array on the side that has the fewer number of elements.
